File: tools/trace.lua
-- Execution tracer for PCSX-Redux.
--
-- Records which of the game's functions actually execute, so decompilation can
-- target code that runs rather than whatever happens to be small or early in
-- address order.  Function choice has so far been driven by size and address,
-- with no evidence about which functions the boot path even calls.
--
-- Reads addresses from build/trace/funcs.txt, sets an execution breakpoint on
-- each, and records the first hit.
--
-- Invoked by tools/trace.py; see there for the required flags (-run, a real
-- BIOS, and stdin held open).

local ffi = require 'ffi'

-- Raw FFI rather than PCSX.addBreakpoint, because that wrapper does an
-- `ffi.cast` per breakpoint and LuaJIT allows only ~1024 live FFI callbacks.
-- 1200+ breakpoints therefore failed with "too many callbacks" -- and because
-- -dofile swallows load errors, it failed completely silently. One shared
-- callback works because the callback already receives the address.
local C = ffi.load 'PCSX'
local BP_EXEC = 0        -- enum BreakpointType { Exec, Read, Write }

local FUNCS_IN = 'build/trace/funcs.txt'
local HITS_OUT = 'build/trace/hits.txt'

local frames = 0
-- From the environment, not a global: main.cc runs every -dofile *before* any
-- -exec, so a global set by -exec is still nil when this script reads it.
local limit = tonumber(os.getenv('TRACE_FRAMES')) or 1800   -- ~30 s at 60 Hz
local hits = {}
local order = {}
local nhits = 0
local handles = {}
local nfuncs = 0
local done = false

local function log(msg)
    -- PCSX.log, not print: print goes to the Lua console, which -no-ui has none
    -- of, so output vanished and the script looked like it had never run.
    PCSX.log('[trace] ' .. msg)
end

-- Must return false, or the emulator pauses on every hit and never advances.
local shared_cb = ffi.cast('bool (*)(uint32_t, unsigned, const char*)',
    function(address, _width, _cause)
        local a = tonumber(address)
        if not hits[a] then
            hits[a] = frames
            order[#order + 1] = a
            nhits = nhits + 1
            local h = handles[a]
            if h ~= nil then
                C.removeBreakpoint(h)      -- coverage is a set; once is enough
                handles[a] = nil
            end
        end
        return false
    end)

local function writeResults(reason)
    if done then return end
    done = true
    local f = io.open(HITS_OUT, 'w')
    if not f then
        log('could not write ' .. HITS_OUT)
        return
    end
    f:write(string.format('# reason=%s frames=%d hit=%d of=%d\n',
                          reason, frames, nhits, nfuncs))
    table.sort(order)                       -- stable output, diffable run to run
    for _, a in ipairs(order) do
        f:write(string.format('%08X %d\n', a, hits[a]))
    end
    f:close()
    log(string.format('%s: %d/%d functions executed over %d frames',
                      reason, nhits, nfuncs, frames))
end

local function armBreakpoints()
    local f = io.open(FUNCS_IN, 'r')
    if not f then
        log('missing ' .. FUNCS_IN .. ' -- run tools/trace.py, not this directly')
        return false
    end
    for line in f:lines() do
        local hex = line:match('^%s*([0-9A-Fa-f]+)')
        if hex then
            local addr = tonumber(hex, 16)
            if addr and not handles[addr] then
                handles[addr] = C.addBreakpoint(addr, BP_EXEC, 4, 'trace',
                                                shared_cb, 'trace')
                nfuncs = nfuncs + 1
            end
        end
    end
    f:close()
    log(string.format('armed %d execution breakpoints with one shared callback',
                      nfuncs))
    return nfuncs > 0
end

if armBreakpoints() then
    PCSX.Events.createEventListener('GPU::Vsync', function()
        frames = frames + 1
        if frames % 200 == 0 then
            log(string.format('frame %d: %d/%d hit', frames, nhits, nfuncs))
        end
        if frames >= limit then
            writeResults('frame-limit')
            PCSX.quit(0)
        end
    end)
    log('tracing until frame ' .. limit)
end

File: tools/trace.py
"""Run the game under PCSX-Redux and record which functions actually execute.

Function choice for decompilation has so far been driven by size and address
order.  That is a poor proxy: a 20-instruction function that never runs on the
way to a duel is worth less than a 200-instruction one that runs every frame.
This produces the coverage set, so effort can follow execution.

PCSX-Redux is used rather than DuckStation because it has a Lua API; DuckStation
has no scripting at all, which is why the existing boot check could only assert
that the process stayed alive.

    py -3 tools/trace.py                    trace from boot, 1800 frames
    py -3 tools/trace.py --frames 600
    py -3 tools/trace.py --report           just re-report the last trace
"""
import argparse
import os
import subprocess
import sys

REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(REPO, "tools"))

import funcs as funcs_mod                                        # noqa: E402

REDUX = os.path.join(REPO, "tools", "bin", "redux", "pcsx-redux.exe")
# The real BIOS, copied from DuckStation. Redux otherwise falls back to
# OpenBIOS, which is a reimplementation and not something to trust for running a
# commercial game -- a trace taken under it would not be evidence about the real
# boot path.
BIOS = os.path.join(REPO, "tools", "bin", "redux", "SCPH1001.BIN")
IMAGE = os.path.join(REPO, "build", "ygofm.bin")
WORK = os.path.join(REPO, "build", "trace")
FUNCS_TXT = os.path.join(WORK, "funcs.txt")
HITS_TXT = os.path.join(WORK, "hits.txt")


def write_function_list():
    """Addresses of every game function, for the Lua side to breakpoint."""
    fns = [f for f in funcs_mod.parse() if f["addr"] < funcs_mod.GAME_END]
    fns.sort(key=lambda f: f["addr"])
    os.makedirs(WORK, exist_ok=True)
    with open(FUNCS_TXT, "w") as fp:
        for f in fns:
            fp.write("%08X %s %d\n" % (f["addr"], f["name"], f["insns"]))
    return {f["addr"]: f for f in fns}


def report(index):
    if not os.path.exists(HITS_TXT):
        print("no trace results at %s" % HITS_TXT)
        return 1

    header, hits = "", {}
    for line in open(HITS_TXT):
        if line.startswith("#"):
            header = line.strip()
            continue
        parts = line.split()
        if len(parts) >= 2:
            hits[int(parts[0], 16)] = int(parts[1])

    print(header)
    total_ins = sum(f["insns"] for f in index.values())
    hit_ins = sum(index[a]["insns"] for a in hits if a in index)
    print("\nexecuted %d of %d functions (%.1f%%)"
          % (len(hits), len(index), 100.0 * len(hits) / max(1, len(index))))
    print("covering %d of %d instructions (%.1f%%)"
          % (hit_ins, total_ins, 100.0 * hit_ins / max(1, total_ins)))

    # The useful output: executed functions, largest first.  These are the ones
    # where decompilation effort buys the most understanding of what the game
    # actually does.
    ranked = sorted((index[a] for a in hits if a in index),
                    key=lambda f: -f["insns"])
    print("\nlargest executed functions (decompile these first):")
    print("  %-20s %-12s %6s  %s" % ("function", "address", "insns", "first seen"))
    for f in ranked[:25]:
        print("  %-20s 0x%08X %6d  frame %d"
              % (f["name"], f["addr"], f["insns"], hits[f["addr"]]))

    never = [f for a, f in index.items() if a not in hits]
    never_ins = sum(f["insns"] for f in never)
    print("\n%d functions never executed (%d instructions, %.1f%% of game code)"
          % (len(never), never_ins, 100.0 * never_ins / max(1, total_ins)))
    return 0


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--frames", type=int, default=1800)
    ap.add_argument("--report", action="store_true")
    ap.add_argument("--timeout", type=int, default=600)
    args = ap.parse_args()

    index = write_function_list()
    if args.report:
        return report(index)

    if not os.path.exists(IMAGE):
        sys.exit("%s missing -- run tools/make_iso.py first" % IMAGE)
    if os.path.exists(HITS_TXT):
        os.remove(HITS_TXT)

    # -run is required: without it the emulator boots, sits paused, and exits.
    # -debugger and -interpreter: the x86-64 dynarec does not check breakpoints,
    # so 1206 armed breakpoints produced zero hits while the game demonstrably
    # ran. The interpreter is slower but is the only mode that observes them.
    cmd = [REDUX, "-no-ui", "-stdout", "-run", "-debugger", "-interpreter",
           "-bios", BIOS,
           "-iso", IMAGE,
           "-dofile", os.path.join("tools", "trace.lua"),
           "-logfile", os.path.join("build", "trace", "redux.log"),
]
    print("tracing %d frames under PCSX-Redux..." % args.frames)
    # stdin must stay open: with -no-ui the TUI reads stdin, and an immediate EOF
    # makes it quit before the game runs at all. That looked exactly like the Lua
    # script failing.
    env = dict(os.environ, TRACE_FRAMES=str(args.frames))
    proc = subprocess.Popen(cmd, cwd=REPO, stdin=subprocess.PIPE, env=env)
    try:
        proc.wait(timeout=args.timeout)
    except subprocess.TimeoutExpired:
        print("emulator hit the %ds timeout; killing it" % args.timeout)
        proc.kill()
        proc.wait(timeout=30)

    return report(index)


if __name__ == "__main__":
    sys.exit(main())
